types/tx.go

```diff
@@ -0,0 +1,102 @@
+package types
+
+import "math/big"
+
+// TxType is the transaction type.
+type TxType byte
+
+// List of supported transaction types
+const (
+	LegacyTx     TxType = 0x0
+	StateTx      TxType = 0x7f
+	DynamicGeeTx TxType = 0x8f
+)
+
+// String returns string representation of the transaction type.
+func (t TxType) String() (s string) {
+	switch t {
+	case LegacyTx:
+		return "LegacyTx"
+	case StateTx:
+		return "StateTx"
+	case DynamicGeeTx:
+		return "DynamicGeeTx"
+	default:
+		return "UnknownTX"
+	}
+}
+
+// Tx represents the general transaction behavior interface.
+// It contains common transaction field getters and setters as
+// well as additional functions for managing transaction.
+// Any modification function should return modified transaction object.
+type Tx interface {
+	// Nonce returns transaction nonce.
+	Nonce() uint64
+
+	// SetNonce sets the given transaction nonce.
+	SetNonce(nonce uint64) Tx
+
+	// GasPrice returns transaction gas price.
+	GasPrice() *big.Int
+
+	// SetGasPrice sets the given transaction gas price.
+	SetGasPrice(gasPrice *big.Int) Tx
+
+	// Gas returns transaction gas.
+	Gas() uint64
+
+	// SetGas sets the given transaction gas.
+	SetGas(gas uint64) Tx
+
+	// To returns transaction receiver address.
+	// It could be nil in case of the state transaction.
+	To() *Address
+
+	// SetTo sets the given transaction receiver address.
+	SetTo(addr Address) Tx
```

Comment:
Setting addr to nil is a requirement for transaction. We can't set addr to nil using this signature.

Reply:
But it is nil by default. Does that not mean if we want to set an address, it should have some value? Could you point me to some place in the code where we explicitly set address to nil please?

Reply:
Well whenever we deploy smart contract, `To` field should be nil, no?

Reply:
Update: https://github.com/0xPolygon/polygon-edge/pull/1111#issuecomment-1380101310